apps/prairielearn/src/lib/preact.tsx

```diff
@@ -0,0 +1,42 @@
+import { type ComponentType, type Attributes } from 'preact';
+import { h } from 'preact';
+import { render } from 'preact-render-to-string';
+
+import { type HtmlSafeString, unsafeHtml } from '@prairielearn/html';
+
+// Based on https://pkg.go.dev/encoding/json#HTMLEscape
+const ENCODE_HTML_RULES: Record<string, string> = {
+  '&': '\\u0026',
+  '>': '\\u003e',
+  '<': '\\u003c',
+  '\u2028': '\\u2028',
+  '\u2029': '\\u2029',
+};
+const MATCH_HTML = /[&><\u2028\u2029]/g;
+
+function escapeJsonForHtml(value: any): string {
+  return JSON.stringify(value).replace(MATCH_HTML, (c) => ENCODE_HTML_RULES[c] || c);
+}
+
+export function renderWithProps<T>(
+  id: string,
+  Component: ComponentType<T>,
+  props: T & Attributes,
+): HtmlSafeString {
```

Comment:
This little function serializes the props so that the client can pick it up, and also renders the component to an HTML string. Currently this is using plain `JSON.stringify`, but it could (and maybe should) use something like https://github.com/flightcontrolhq/superjson if we want to actually ship this as a core little primitive.

Reply:
Is there some way with `superjson` to use `zod` or something like that to get back strongly-typed objects like our `db-types` objects?

Reply:
We'd rely on `superjson` to transmit things exactly as they are, so we wouldn't need Zod on the client. `renderWithProps` already does all the type-checking necessary to ensure that the props match what the component expects. If `superjson` serializes and deserializes precisely, we'll get exactly the same (already-type-checked) data we had on the server.